runtimevar/runtimevar_test.go

```diff
@@ -18,16 +18,183 @@ package runtimevar_test
+	w.calls = w.calls[1:]
+	if prevErr != nil && c.err != nil && c.err.Error() == prevErr.Error() {
+		w.t.Log("  --> same error")
+		return nil, nil, 0, nil
+	}
+	if prevVersion != nil && prevVersion == c.version {
+		w.t.Log("  --> same version")
+		return nil, nil, 0, nil
+	}
+	w.t.Logf("  --> returning val %v, version %v, err %v", c.v, c.version, c.err)
+	return c.v, c.version, 0, c.err
+}
+
+// watchResp encapsulates the expected result of a Watch call.
+type watchResp struct {
+	snap runtimevar.Snapshot
+	err  bool
+}
+
+func TestVariable(t *testing.T) {
+
+	fail1, fail2 := errors.New("fail1"), errors.New("fail2")
+	var1 := &driver.Variable{Value: 42, UpdateTime: time.Now()}
+	var2 := &driver.Variable{Value: 43, UpdateTime: time.Now().Add(1 * time.Minute)}
+	snap1 := runtimevar.Snapshot{Value: var1.Value, UpdateTime: var1.UpdateTime}
+	snap2 := runtimevar.Snapshot{Value: var2.Value, UpdateTime: var2.UpdateTime}
+
+	tests := []struct {
+		name    string
+		doneCtx bool
+		calls   []*watchVariableCall
+		// Watch will be called once for each entry in want.
+		want []*watchResp
+	}{
+		{
+			name:    "Done context returns error with no WatchVariable calls",
+			doneCtx: true,
+			calls:   nil,
+			want: []*watchResp{
+				&watchResp{err: true},
+				&watchResp{err: true},
+			},
+		},
+		{
+			name: "Repeated errors don't return until it changes to a different error",
+			calls: []*watchVariableCall{
+				&watchVariableCall{err: fail1},
+				&watchVariableCall{err: fail1},
+				&watchVariableCall{err: fail1},
+				&watchVariableCall{err: fail1},
+				&watchVariableCall{err: fail2},
+			},
+			want: []*watchResp{
+				&watchResp{err: true},
+				&watchResp{err: true},
+			},
+		},
+		{
+			name: "Repeated errors don't return until it changes to a value",
+			calls: []*watchVariableCall{
+				&watchVariableCall{err: fail1},
+				&watchVariableCall{err: fail1},
+				&watchVariableCall{err: fail1},
+				&watchVariableCall{err: fail1},
+				&watchVariableCall{v: var1},
+			},
+			want: []*watchResp{
+				&watchResp{err: true},
+				&watchResp{snap: snap1},
+			},
+		},
+		{
+			name: "Repeated values don't return until it changes to an error",
+			calls: []*watchVariableCall{
+				&watchVariableCall{v: var1, version: 1},
+				&watchVariableCall{v: var1, version: 1},
+				&watchVariableCall{v: var1, version: 1},
+				&watchVariableCall{v: var1, version: 1},
+				&watchVariableCall{err: fail1},
+			},
+			want: []*watchResp{
+				&watchResp{snap: snap1},
+				&watchResp{err: true},
+			},
+		},
+		{
+			name: "Repeated values don't return until it changes to a different version",
+			calls: []*watchVariableCall{
+				&watchVariableCall{v: var1, version: 1},
+				&watchVariableCall{v: var1, version: 1},
+				&watchVariableCall{v: var1, version: 1},
+				&watchVariableCall{v: var1, version: 1},
+				&watchVariableCall{v: var2, version: 2},
+			},
+			want: []*watchResp{
+				&watchResp{snap: snap1},
+				&watchResp{snap: snap2},
+			},
+		},
```

Comment:
I could add a test that returns a non-zero waitTime and verify that the next WatchVariable call is at least <waitTime> after the previous one. WDYT?

Reply:
I don't understand what the test is. Is it that the amount of time has elapsed? If that's the case, I don't think that's needed. (See my comment above about the duration being a hint.)